### PR TITLE
add winloop for windows if use_uvloop is used on windows

### PR DESCRIPTION
**Pull request by Vizonex**
## Changes

Fixes #959 

This attempts to add winloop support to anyio. And is my second attempt. This time I not hook pre-commit so that stressful things don't happen again. I am going to be doing these steps one at a time so that I can be guided in the right direction instead of screwing up.

## Checklist

If this is a user-facing code change, like a bugfix or a new feature, please ensure that
you've fulfilled the following conditions (where applicable):

- [x] You've added tests (in `tests/`) added which would fail without your patch
- [x] You've updated the documentation (in `docs/`, in case of behavior changes or new
features)
- [x] You've added a new changelog entry (in `docs/versionhistory.rst`).

If this is a trivial change, like a typo fix or a code reformatting, then you can ignore
these instructions.

### Updating the changelog

If there are no entries after the last release, use `**UNRELEASED**` as the version.
If, say, your patch fixes issue <span>#</span>123, the entry should look like this:

```
- Fix big bad boo-boo in task groups
  (`#123 <https://github.com/agronholm/anyio/issues/123>`_; PR by @yourgithubaccount)
```

If there's no issue linked, just link to your pull request instead by updating the
changelog after you've created the PR.

### CODE REVIEW

**Comment by Vizonex** [comment]
@agronholm I'll make sure to grab what I previously had for documentation about winloop real quick.

File: tests/conftest.py [hunk 1]
```diff
@@ -39,6 +53,11 @@
         marks=uvloop_marks,
         id="asyncio+uvloop",
     ),
+    pytest.param(
+        ("winloop", {"debug": True, "loop_factory": winloop.new_event_loop}),
```
Comment (agronholm) [inline tests/conftest.py]: There's no backend named "winloop".
```suggestion
        ("asyncio", {"debug": True, "loop_factory": winloop.new_event_loop}),
```

File: tests/conftest.py [hunk 2]
```diff
@@ -30,6 +30,20 @@
...
+except ImportError:
+    winloop_marks.append(pytest.mark.skip(reason="winloop not available"))
+    winloop = Mock()
+else:
+    if hasattr(asyncio.AbstractEventLoop, "shutdown_default_executor") and not hasattr(
+        winloop.loop.Loop, "shutdown_default_executor"
+    ):
+        winloop_marks.append(
+            pytest.mark.skip(reason="winloop is missing shutdown_default_executor()")
+        )
```
Comment (agronholm) [inline tests/conftest.py]: Does `winloop` have releases that lack `shutdown_default_executor()`?

Comment (Vizonex) [reply tests/conftest.py]: Actually I don't think it does. Let me remove this

File: tests/test_sockets.py [hunk 3]
```diff
@@ -2195,6 +2195,7 @@ async def test_getaddrinfo() -> None:
 
 
 @pytest.mark.parametrize("sock_type", [socket.SOCK_STREAM, socket.SOCK_STREAM])
+@pytest.mark.skipif(sys.platform == "win32", reason="getaddrinfo can be incorrect on winloop")
```
Comment (agronholm) [inline tests/test_sockets.py]: Only skip it on winloop, not on vanilla asyncio

Comment (agronholm) [reply tests/test_sockets.py]: Better yet, fix the test logic to check for the correct values

Comment (Vizonex) [reply tests/test_sockets.py]: Oh ok my bad Maybe I should make a wrapper that skips if on winloop then?

Comment (agronholm) [reply tests/test_sockets.py]: Why can't you just fix the logic on line 2205?

File: tests/test_subprocesses.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
 pytestmark = pytest.mark.anyio
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="winloop fails this shell test")
```
Comment (agronholm) [inline tests/test_subprocesses.py]: You shouldn't skip this universally on Windows if it only fails on winloop

File: tests/test_subprocesses.py [hunk 5]
```diff
@@ -275,6 +276,9 @@ async def test_process_aexit_cancellation_closes_standard_streams(
         await process.stderr.receive(1)
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32", reason="winloop doesn't have have umask or special kwargs"
+)
```
Comment (agronholm) [inline tests/test_subprocesses.py]: As above, no skipping this on all loop implementations

File: tests/test_sockets.py [hunk 6]
```diff
@@ -2195,11 +2197,9 @@ async def test_getaddrinfo() -> None:
 
 
 @pytest.mark.parametrize("sock_type", [socket.SOCK_STREAM, socket.SOCK_STREAM])
-@pytest.mark.skipif(
-    sys.platform == "win32", reason="getaddrinfo can be incorrect on winloop"
-)
 async def test_getaddrinfo_ipv6addr(
     sock_type: Literal[socket.SocketKind.SOCK_STREAM],
+    skip_if_winloop
```
Comment (agronholm) [inline tests/test_sockets.py]: You could've just followed the existing examples of skipping when uvloop is detected, yes?

File: tests/test_sockets.py [hunk 7]
```diff
@@ -2196,7 +2196,7 @@ async def test_getaddrinfo() -> None:
 
 @pytest.mark.parametrize("sock_type", [socket.SOCK_STREAM, socket.SOCK_STREAM])
 async def test_getaddrinfo_ipv6addr(
-    sock_type: Literal[socket.SocketKind.SOCK_STREAM],
+    sock_type: Literal[socket.SocketKind.SOCK_STREAM], skip_if_winloop
 ) -> None:
     # IDNA trips up over raw IPv6 addresses
     proto = 0 if platform.system() == "Windows" else 6
```
Comment (agronholm) [inline tests/test_sockets.py]: Winloop does this correctly so the expected value should be 6

File: tests/test_subprocesses.py [hunk 8]
```diff
@@ -309,6 +312,7 @@ async def test_py39_arguments(
     argvalue_factory: Callable[[], Any],
     anyio_backend_name: str,
     anyio_backend_options: dict[str, Any],
+    skip_if_winloop: None,
```
Comment (agronholm) [inline tests/test_subprocesses.py]: Instead of this, check for winloop like we're already checking for uvloop:
`and anyio_backend_options["loop_factory"].__module__ in ("uvloop", "winloop")`

**Comment by agronholm** [comment]
I will approve this once you change the test skipping to use the same method as with uvloop. It's okay to add a utility function in the test suite to check for winloop/uvloop as you would now need to duplicate the same nontrivial logic.

File: tests/test_subprocesses.py [hunk 9]
```diff
@@ -42,6 +42,11 @@
 async def test_run_process(
     shell: bool, command: str | list[str], anyio_backend_name: str
 ) -> None:
+    if anyio_backend_name == "asyncio" and sys.platform == "win32":
+        import asyncio
+
+        if asyncio.get_event_loop().__module__ == "winloop":
+            pytest.skip(reason="winloop produces a non-zero exit with status 1")
```
Comment (agronholm) [inline tests/test_subprocesses.py]: Why would winloop fail with this basic functionality?

**Comment by Vizonex** [comment]
is there anything left or is this mergeable?

**Comment by Vizonex** [comment]
> Why is there an added winloop skip in `test_run_process()`? It makes no sense. Does running subprocesses really not work on winloop?
there seems to be a bug that doesn't want it to pass and seems to be a problem with the library itself maybe. I'll see about making a duplicate of the test but basically the one that is failing is tests/test_subprocesses.py::test_run_process[asyncio+uvloop-shell] for some strange reason. I'll see about patching it in a later update.

**Comment by agronholm** [comment]
> > Why is there an added winloop skip in `test_run_process()`? It makes no sense. Does running subprocesses really not work on winloop?
> 
> @agronholm there seems to be a bug that doesn't want it to pass and seems to be a problem with the library itself maybe. I'll see about making a duplicate of the test but basically the one that is failing is tests/test_subprocesses.py::test_run_process[asyncio+uvloop-shell] for some strange reason. I'll see about patching it in a later update.

I'd really like to see this fixed in winloop before this PR is merged. The next release of AnyIO is some time away given how v4.10.0 was just released, so there's plenty of time.

**Review by agronholm — APPROVED** [review]

**Comment by Vizonex** [comment]
@agronholm Looks like my long 2 month journey to get this in is finally finished. What a relief. I'll be able to focus on some of my other goals now thanks to this being completed.

**Comment by agronholm** [comment]
Thanks!
